gcs/src/gcs_group.cpp

```diff
@@ -60,6 +61,9 @@ gcs_group_init (gcs_group_t* group, gcache_t* const cache,
 
     group->last_applied_proto_ver = -1;
 
+    gu_mutex_init (&group->index_lock, NULL);
+    gu_cond_init (&group->index_cond, NULL);
```

Comment:
If we relook at places where group->my_idx is modified then it is fortunately only a single place "gcs_group_handle_comp_msg".

Value of group->my_idx is modified from
VALID -> -1 -> VALID.

This means we simply needs a mutex protector.

pthread_mutex_lock(x)
...  group_nodes_free (group);

    group->my_idx = new_my_idx;
    group->num    = new_nodes_num;
    group->nodes  = new_nodes;

pthread_mutex_unlock(mutex)

----------

If show status is called before group_nodes_frees set it to -1 then show status will get VALID value and group_nodes_free will not proceed till show status is complete due to mutex lock.
If group_nodes_free is in progress to set group->my_idx = -1 then due to mutex lock show status will wait.